src/luarocks/manif/writer.lua

```diff
@@ -286,73 +313,105 @@ function writer.make_manifest(repo, deps_mode, remote)
-      end
-   end
-   local ok, err = store_results(results, manifest, dep_handler)
+   local ok, err = store_results(results, manifest)
    if not ok then return nil, err end
 
    if remote then
       local cache = {}
       for luaver in util.lua_versions() do
          local vmanifest = { repository = {}, modules = {}, commands = {} }
-         local dep_handler = function(manifest)
-            filter_by_lua_version(manifest, luaver, repo, cache)
-         end
-         store_results(results, vmanifest, dep_handler)
+         local ok, err = store_results(results, vmanifest)
+         filter_by_lua_version(vmanifest, luaver, repo, cache)
          save_table(repo, "manifest-"..luaver, vmanifest)
       end
+   else
+      update_dependencies(manifest, deps_mode)
    end
 
    return save_table(repo, "manifest", manifest)
 end
 
---- Load a manifest file from a local repository and add to the repository
--- information with regard to the given name and version.
--- A file called 'manifest' will be written in the root of the given
--- repository directory.
+--- Update manifest file for a local repository
+-- adding information about a version of a package installed in that repository.
 -- @param name string: Name of a package from the repository.
 -- @param version string: Version of a package from the repository.
 -- @param repo string or nil: Pathname of a local repository. If not given,
 -- the default local repository is used.
 -- @param deps_mode string: Dependency mode: "one" for the current default tree,
 -- "all" for all trees, "order" for all trees with priority >= the current default,
 -- "none" for using the default dependency mode from the configuration.
--- @return boolean or (nil, string): True if manifest was generated,
+-- @return boolean or (nil, string): True if manifest was updated successfully,
 -- or nil and an error message.
-function writer.update_manifest(name, version, repo, deps_mode)
+function writer.add_to_manifest(name, version, repo, deps_mode)
    assert(type(name) == "string")
    assert(type(version) == "string")
-   repo = path.rocks_dir(repo or cfg.root_dir)
+   local rocks_dir = path.rocks_dir(repo or cfg.root_dir)
    assert(type(deps_mode) == "string")
-   
+
    if deps_mode == "none" then deps_mode = cfg.deps_mode end
 
-   local manifest, err = manif.load_manifest(repo)
+   local manifest, err = manif.load_local_manifest(rocks_dir)
    if not manifest then
       util.printerr("No existing manifest. Attempting to rebuild...")
-      local ok, err = writer.make_manifest(repo, deps_mode)
-      if not ok then
-         return nil, err
-      end
-      manifest, err = manif.load_manifest(repo)
-      if not manifest then
-         return nil, err
-      end
+      -- Manifest built by `manif.make_manifest` should already
+      -- include information about given name and version,
+      -- no need to update it.
+      return manif.make_manifest(rocks_dir, deps_mode)
```

Comment:
This should refer to `writer.make_manifest` now, same for `remove_from_manifest`